src/_pytest/main.py

```diff
@@ -575,6 +577,10 @@ def _collect(self, arg):
             yield from m
 
     def _collectfile(self, path, handle_dupes=True):
+        key = (path, handle_dupes)
+        assert key not in self._collected, key
```

Comment:
i'm not familiar with the code, how does this interact with handle_dupes?

Reply:
`handle_duples` is also a hack on top (IIRC even by me already).  I just added it here.  The code is not meant to stay.
`_collectfile` could use a cache/lookup to work around this - the main issue is that hooks shouldn't really be called multiple times.  I've only noticed this when debugging some `--lf` stuff.

Reply:
https://github.com/pytest-dev/pytest/pull/4319 . i actually merged it

i believe this will b better if we shift packages around so they are parents of modules and not alongside of modules
(then we can also attach conftests to packages/blank packages and finally get rid of the nodeid string matching hell)

Reply:
Sounds good!